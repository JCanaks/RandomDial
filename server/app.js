import express from 'express';
import cors from 'cors';
import morgan from 'morgan';
import 'dotenv/config';
import bodyParser from 'body-parser';
import routes from './routes';

const app = express();
app.use(cors());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());
app.use(morgan('dev'));
app.use('/api', routes);

app.get('/', (req, res) => {
  res.send({
    message: 'Welcome to RandomDail, your Simple Random Phone Number Generator',
  });
});

app.use((err, req, res, next) => res.status.json({
  message: 'Internal Server error',
}));

export default app;
