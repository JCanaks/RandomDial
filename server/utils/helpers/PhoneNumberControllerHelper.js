/* eslint-disable no-param-reassign */
import readline from 'readline';
import path from 'path';
import fs from 'fs';

export const generateReadlineInstance = () => {
  const filePath = '../../database/storage/phoneNumbers.txt';
  const readLineInstance = readline.createInterface({
    input: fs.createReadStream(path.join(__dirname, filePath),
      { flags: 'a+' }),
  });
  return readLineInstance;
};

export const getPhoneNumberList = (readLineInstance, phoneBook) => {
  try {
    readLineInstance.on('line', (line) => {
      // eslint-disable-next-line no-param-reassign
      phoneBook.total += 1;
      phoneBook.phoneNumbers.push(line);
    });
  } catch (error) {
    throw error;
  }
};

export const responseFormat = (message, data = {}) => ({
  message,
  data,
});

export const generateRandomPhoneNumbers = (amount) => {
  const phoneNumbers = [];
  for (let i = 0; i < amount; i += 1) {
    const randomPhoneNumber = `0${Math.floor(Math.random() * 900000000) + 100000000}`;
    phoneNumbers.push(randomPhoneNumber);
  }
  const generatedDetails = {
    total: phoneNumbers.length,
    phoneNumbers,
  };
  return generatedDetails;
};

export const storeRandomPhoneNumbers = (phoneNumbers) => {
  const filePath = '../../database/storage/phoneNumbers.txt';
  const writeStream = fs.createWriteStream(path.join(__dirname, filePath),
    { flags: 'a+' });
  phoneNumbers.forEach((element) => {
    writeStream.write(`${element}\n`);
  });
  writeStream.end();
};

export const getMaxAndMinNumbers = (readLineInstance, phoneBook) => {
  try {
    readLineInstance.on('line', (number) => {
      const phoneNumber = parseInt(number, 10);

      // When there is no phone number in the file storage
      // eslint-disable-next-line no-restricted-globals
      if (isNaN(phoneNumber)) {
        return;
      }

      if (phoneBook.total === 0) {
        phoneBook.min = number;
      }
      phoneBook.min = phoneNumber < parseInt(phoneBook.min, 10) ? number : phoneBook.min;
      phoneBook.max = phoneNumber > parseInt(phoneBook.max, 10) ? number : phoneBook.max;
      phoneBook.total += 1;
      phoneBook.phoneNumbers.push(number);
    });
  } catch (error) {
    throw error;
  }
};

export const sortAscending = (phoneNumbers) => {
  phoneNumbers.sort((a, b) => a - b);
};

export const sortDescending = (phoneNumbers) => {
  phoneNumbers.sort((a, b) => b - a);
};

export const clearFile = () => {
  const filePath = '../../database/storage/phoneNumbers.txt';
  fs.writeFile(path.join(__dirname, filePath), '', { flag: 'w+' }, (err) => {
    if (err) throw err;
  });
};
