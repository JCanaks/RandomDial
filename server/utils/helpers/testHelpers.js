import fs from 'fs';
import path from 'path';

export const seedNumbers = () => {
  const mockSeedData = '0333333333\n0222222222\n0111111111\n0555555555\n0444444444';
  const filePath = '../../database/storage/phoneNumbers_test.txt';
  fs.writeFile(path.join(__dirname, filePath), mockSeedData, { flag: 'w+' });
};

export default seedNumbers;
