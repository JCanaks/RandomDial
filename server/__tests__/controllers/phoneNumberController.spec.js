import '@babel/polyfill';
import request from 'supertest';
import app from '../../app';
import { clearFile } from '../../utils/helpers/PhoneNumberControllerHelper';
import seedNumbers from '../../utils/helpers/testHelpers';

describe('Testing the PhoneNumber Controller', () => {
    beforeAll(async () => {
        await clearFile();
        await seedNumbers();
      });
      afterAll(async () => {
        await clearFile();
      });
    test('Get /phonenumber - Getting the list of phone numbers', async () => {
        const response = await request(app).get('/api/phonenumbers');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(200);
    });

    test('Get /sort - sorting the list of phone numbers in ascending order and finding max/min numbers', async () => {
        const response = await request(app).get('/api/sort?order=asc');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(200);
    })
    test('Get /sort - sorting the list of phone numbers in descending order and finding max/min numbers', async () => {
        const response = await request(app).get('/api/sort?order=desc');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(200);
    })
    test('Get /sort - sorting the list of phone numbers without order and finding max/min numbers', async () => {
        const response = await request(app).get('/api/sort');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(200);
    });

    test('POST /phonenumbers - generating random phone numbers', async () => {
        const requestBody = {
            amount: 10
        }
        const response = await request(app).post('/api/phonenumbers').send(requestBody);
        expect(response.body.message).toEqual(`${requestBody.amount} phone number(s) successfully generated`);
        expect(response.body.data.total).toEqual(requestBody.amount);
        expect(response.body.data.phoneNumbers.length).toEqual(requestBody.amount);
        expect(response.statusCode).toBe(201);
    });

    test('POST /phonenumbers - generating random phone numbers above 500', async () => {
        const requestBody = {
            amount: 1000
        }
        const response = await request(app).post('/api/phonenumbers').send(requestBody);
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(400);
    });

    test('POST /phonenumbers - generating random phone numbers with no amount specified', async () => {
        const response = await request(app).post('/api/phonenumbers');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(400);
    });

    test('Get /clear - clearing the list of phone numbers', async () => {
        const response = await request(app).delete('/api/clear');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(200);
    });

    test('Get /phonenumber - Getting an empty list of phone numbers', async () => {
        const response = await request(app).get('/api/phonenumbers');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(200);
    });
    test('Get /sort - sorting an empty list of phone numbers', async () => {
        const response = await request(app).get('/api/sort');
        expect(response.body).toMatchSnapshot();
        expect(response.statusCode).toBe(200);
    });
});




