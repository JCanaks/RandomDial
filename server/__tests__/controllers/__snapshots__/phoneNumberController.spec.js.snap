// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the PhoneNumber Controller Get /clear - clearing the list of phone numbers 1`] = `
Object {
  "data": Object {},
  "message": "Phone Numbers Successfully deleted",
}
`;

exports[`Testing the PhoneNumber Controller Get /phonenumber - Getting an empty list of phone numbers 1`] = `
Object {
  "data": Object {
    "phoneNumbers": Array [],
    "total": 0,
  },
  "message": "No phone numbers found. Pls generate new phone numbers",
}
`;

exports[`Testing the PhoneNumber Controller Get /phonenumber - Getting the list of phone numbers 1`] = `
Object {
  "data": Object {
    "phoneNumbers": Array [
      "0333333333",
      "0222222222",
      "0111111111",
      "0555555555",
      "0444444444",
    ],
    "total": 5,
  },
  "message": "Phone Numbers Fetched Successfully",
}
`;

exports[`Testing the PhoneNumber Controller Get /sort - sorting an empty list of phone numbers 1`] = `
Object {
  "data": Object {
    "phoneNumbers": Array [],
    "total": 0,
  },
  "message": "No phone numbers found. Pls generate new phone numbers",
}
`;

exports[`Testing the PhoneNumber Controller Get /sort - sorting the list of phone numbers in ascending order and finding max/min numbers 1`] = `
Object {
  "data": Object {
    "max": "0555555555",
    "min": "0111111111",
    "phoneNumbers": Array [
      "0111111111",
      "0222222222",
      "0333333333",
      "0444444444",
      "0555555555",
    ],
    "total": 5,
  },
  "message": "Phone Numbers Fetched Successfully",
}
`;

exports[`Testing the PhoneNumber Controller Get /sort - sorting the list of phone numbers in descending order and finding max/min numbers 1`] = `
Object {
  "data": Object {
    "max": "0555555555",
    "min": "0111111111",
    "phoneNumbers": Array [
      "0555555555",
      "0444444444",
      "0333333333",
      "0222222222",
      "0111111111",
    ],
    "total": 5,
  },
  "message": "Phone Numbers Fetched Successfully",
}
`;

exports[`Testing the PhoneNumber Controller Get /sort - sorting the list of phone numbers without order and finding max/min numbers 1`] = `
Object {
  "data": Object {
    "max": "0555555555",
    "min": "0111111111",
    "phoneNumbers": Array [
      "0333333333",
      "0222222222",
      "0111111111",
      "0555555555",
      "0444444444",
    ],
    "total": 5,
  },
  "message": "Phone Numbers Fetched Successfully",
}
`;

exports[`Testing the PhoneNumber Controller POST /phonenumbers - generating random phone numbers above 500 1`] = `
Object {
  "data": Array [],
  "message": "Cannot Generate more than 500 numbers at a time",
}
`;

exports[`Testing the PhoneNumber Controller POST /phonenumbers - generating random phone numbers with no amount specified 1`] = `
Object {
  "data": Array [],
  "message": "No number(s) generated please specify an amount (integer) greater than or equal to 1",
}
`;
